kongswap_adaptor/tests/lifecycle.rs

```diff
@@ -0,0 +1,712 @@
+    // where amount_sns = 1 * E8 - 2 * FEE_SNS
+    // SNS:
+    //      reserve_sns += amount_sns
+    // ICP:
+    //      reserve_icp -= amount_icp
+    //      lp_fee_1 = 30 * amount_icp / 10_000
+    trade(
+        &mut agent,
+        sns_ledger_canister_id,
+        icp_ledger_canister_id,
+        1 * E8,
+        true,
+    )
+    .await;
+    // Phase IV: another deposit
+    // Now, as the pool is no longer in 1:1 price point, and we have more SNS
+    // some amount of ICP will be returned to the user (one more fee)
+    // The amount of icp to be moved into the pool is calculated as
+    // amount_icp = (reserve_icp / reserve_sns) * deposit_amount
+    // where depoit_amount = 10 * E8 - 2 * FEE_SNS
+    // SNS:
+    //      reserve_0 += deposit_amount
+    // ICP:
+    //      reserve_1 += amount_icp calculated above
+    deposit(
+        &mut agent,
+        sns_ledger_canister_id,
+        icp_ledger_canister_id,
+        kong_adaptor_canister_id,
+        treasury_sns_account,
+        treasury_icp_account,
+        10 * E8,
+    )
+    .await;
+    // Phase V: second trade
+    // Here, we try to swap 1 * E8 of ICP and receive SNS in return.
+    // The amount of SNS withdrawn from the kongswap backend is
+    // amount_sns = amount_icp * reserve_sns / (reserve_icp + amount_icp)
+    // where amount_icp = 1 * E8 - 2 * FEE_ICP
+    // SNS:
+    //      reserve_sns= 121 * E8 - 8 * FEE_SNS - amount_sns
+    //      lp_fee_0 = 30 * amount_sns / 10_000
+    // ICP:
+    //      reserve_icp += E8
+    //      lp_fee_1 = 30 * amount_icp / 10_000
+    trade(
+        &mut agent,
+        sns_ledger_canister_id,
+        icp_ledger_canister_id,
+        1 * E8,
+        false,
+    )
+    .await;
+
+    // VI: final phase: withdrawal
+    {
+        let (balances_0, balances_1) = withdraw(&mut agent, kong_adaptor_canister_id).await;
+        assert_eq!(
+            balances_0.fee_collector.as_ref().unwrap().amount_decimals,
+            Nat::from(8 * FEE_SNS)
+        );
+        assert_eq!(
+            balances_1.fee_collector.as_ref().unwrap().amount_decimals,
+            Nat::from(9 * FEE_ICP)
+        );
+
+        assert_eq!(
+            balances_0
+                .external_custodian
+                .as_ref()
+                .unwrap()
+                .amount_decimals,
+            Nat::from(0_u64),
+            "There should be no SNS left in the DEX"
+        );
+        assert_eq!(
+            balances_1
+                .external_custodian
+                .as_ref()
+                .unwrap()
+                .amount_decimals,
+            Nat::from(0_u64),
+            "There should be no ICP left in the DEX"
+        );
+
+        assert_eq!(
+            balances_0.treasury_owner.as_ref().unwrap().amount_decimals,
+            Nat::from(11999249291_u64)
+        );
+        assert_eq!(
+            balances_1.treasury_owner.as_ref().unwrap().amount_decimals,
+            Nat::from(12001107784_u64)
+        );
+    }
```

Comment:
Should we also have an assertion for some other parties' balances, e.g., the suspense is hopefully very low?